Products/ZenModel/actions.py

```diff
@@ -394,76 +395,110 @@ def _encodeBody(self, body):
             plain_body = MIMEText(body.decode('ascii', 'ignore'))
         return plain_body
 
+    def _targetsByTz(self, dmd, targets):
+        """
+        Take timezone from user property to convert a event time in
+        notification and also group targets emails by those timezones.
+        """
+        targets_by_timezones = {}
+        for user in dmd.ZenUsers.getAllUserSettings():
+            if user.email in targets:
+                if targets_by_timezones.get(user.timezone):
+                    targets_by_timezones[user.timezone].add(user.email)
+                else:
+                    targets_by_timezones[user.timezone] = set([user.email])
+        return targets_by_timezones
+
+    def _adjustToTimezone(self, millis, timezone):
+        """
+        Convert event timestamp to timezone from user property.
+        """
+        return long(isoToTimestamp(convertTimestampToTimeZone(
+                    millis / 1000, timezone, "%Y-%m-%d %H:%M:%S"
+                    ))) * 1000
+
     def executeBatch(self, notification, signal, targets):
         log.debug("Executing %s action for targets: %s", self.name, targets)
         self.setupAction(notification.dmd)
+        targets_by_timezones = self._targetsByTz(notification.dmd, targets)
+        original_lst = signal.event.last_seen_time
+        original_fst = signal.event.first_seen_time
+        original_sct = signal.event.status_change_time
+        for target_timezone, targets in targets_by_timezones.iteritems():
+            # Convert timestamp to user timezone
+            signal.event.last_seen_time = self._adjustToTimezone(
+                original_lst, target_timezone)
+            signal.event.first_seen_time = self._adjustToTimezone(
+                original_fst, target_timezone)
+            signal.event.status_change_time =  self._adjustToTimezone(
+                original_sct, target_timezone)
```

Comment:
The use of those `original_lst`, `original_fst`, `and original_sct` is that after first iteration in `for` I will convert the `signal.event.last_seen_time` so it will be a timestamp in new timezone, in the next iteration I will work with it again, the one in a timezone, so if I will move from it I will get a bad time for other timezones.
For converting to different timezones timestamp has to be in UTC.